transport-msg.tex

```diff
@@ -437,3 +429,249 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+VIRTIO_MSG_TYPE and have the following format:
+
+\begin{tabularx}{\textwidth}{|l|l|l|X|}
+\hline
+Name & Offset & Size (bytes) & Content \\
+\hline \hline
+VIRTIO_MSG_TYPE & 0 & 1 & Bit[0]: 0=Request, 1=Answer \newline Bit[1]: 1=bus-msg \newline Bit[2-7]: Reserved \\
+\hline
+VIRTIO_MSG_ID & 1 & 1 & Message ID \\
+\hline
+VIRTIO_MSG_PAYLOAD & 2 & 38 & Payload \\
+\hline
+\end{tabularx}
+
+\subsubsection{Message Definition}\label{sec:Virtio Transport Options / Virtio Over Messages / Message Bus / Message Definition}
+
+The following table is listing the different Virtio-MSG Bus messages and the
+sender for each of them.
+
+\begin{tabular}{|l|l|l|}
+\hline
+Name & ID & Sender \\
+\hline
+\hline
+Reserved                    & 0x0  &        \\
+\hline
+BUS_MSG_ERROR               & 0x1  & Any    \\
+\hline
+BUS_MSG_GET_DEVICES         & 0x2  & Driver \\
+\hline
+BUS_MSG_DEVICE_ADDED        & 0x3  & Device \\
+\hline
+BUS_MSG_DEVICE_REMOVED      & 0x4  & Device \\
+\hline
+BUS_MSG_PING                & 0x5  & Any    \\
+\hline
+BUS_MSG_RESET               & 0x6  & Driver  \\
+\hline
+\end{tabular}
+
+Bus message ID 6-127 are reserved for future use in the specification and
+message ID 128-255 can be used for implementation specific needs.
+
+The following sections are giving more details of the usage of each message
+and the encoding of the payload for the request and answer (when applicable).
+
+\newcommand{\busdef}[1]{\paragraph{BUS_MSG_#1}\label{sec:Virtio Transport Options / Virtio Over Messages / Message Bus / Message Definition / BUS_MSG_#1}}
+
+\busdef{ERROR}
+
+The Error message is used by one side of the bus to inform the other side that
+one message could not be processed correctly. The message contains an error
+code defined using the Standard Posix Error Codes (TODO: add a reference).
+
+This message shall be used for Bus errors not specific to a particular device.
+In case of a device specific error not impacting the complete bus, the
+VIRTIO_MSG_ERROR shall be used.
+
+As for the VIRTIO_MSG_ERROR, this message supports to have a NULL terminating
+String that can be used to give more information printable to the user.
+
+This message can be received by the frontend or the backend side and can be
+send as answer to an other message or directly to signal an error.
+
+\begin{tabular}{|l|l|l|l|}
+\hline
+Type & Offset & Size (bytes) & Content \\
+\hline \hline
+Any     & 0 & 4 & Posix Error Code \\
+        & 4 & 1 & Message ID which generated an error or 0 if none \\
+        & 5 & 1 & Data Type: 0 (Reserved), 1 (STRING), 2-255 (Implementation Defined) \\
+        & 6 & 32 & Null Terminated String if Data Type is STRING else Implementation Defined \\
+\hline
+\end{tabular}
+
+\busdef{GET_DEVICES}
+
+The Get Devices message is used by the frontend to retrieve the ID of the
+devices available on the backend of the bus.
+
+The frontend requests the availibility at an offset and the backend returns a
```

Comment:
should we use  "word selection" instead of "offset" term as done for the virtio MMIO? ( see GuestFeaturesSel as example)

Reply:
I will rephrase

Reply:
@arnopo word selection in this context looks a bit weird. The FeatureSel is actually selecting a "Word" (ie 32bis) using a different register (ie you write in feature sel to configure where you read using deviceFeatures) where here we have send a message to request a list of devices availability.
I am thinking of changing the sentence this way:
The frontend requests the availibility of a group of 256 device IDs and the
backend returns a bitmap indicating which device IDs are available at the
requested device ID offset.
Tell me what you think.